src/java.base/share/classes/jdk/internal/foreign/BufferStack.java

```diff
@@ -0,0 +1,188 @@
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.foreign;
+
+import jdk.internal.misc.CarrierThreadLocal;
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.SegmentAllocator;
+import java.lang.ref.Reference;
+import java.util.Objects;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static java.lang.foreign.ValueLayout.JAVA_LONG;
+
+/**
+ * A buffer stack that allows efficient reuse of memory segments. This is useful in cases
+ * where temporary memory is needed.
+ * <p>
+ * Use the factory {@link #of(long)} to create new instances of this class.
+ * <p>
+ * Note: The reused segments are neither zeroed out before nor after re-use.
+ */
+public record BufferStack(long size, CarrierThreadLocal<PerThread> tl) {
+
+    /**
+     * {@return a new Arena that tries to provide {@code size} and {@code byteAlignment}
+     *          allocations by recycling the BufferStacks internal memory}
+     *
+     * @param size          to be reserved from this BufferStacks internal memory
+     * @param byteAlignment to be used for reservation
+     */
+    @ForceInline
+    public Arena pushFrame(long size, long byteAlignment) {
+        return tl.get().pushFrame(size, byteAlignment);
+    }
+
+    /**
+     * {@return a new Arena that tries to provide {@code layout}
+     *          allocations by recycling the BufferStacks internal memory}
+     *
+     * @param layout for which to reserve internal memory
+     */
+    @ForceInline
+    public Arena pushFrame(MemoryLayout layout) {
+        return pushFrame(layout.byteSize(), layout.byteAlignment());
+    }
+
+    @Override
+    public String toString() {
+        return "BufferStack[" + size + "]";
+    }
+
+    @Override
+    public int hashCode() {
+        return System.identityHashCode(this);
```

Comment:
What's the value in these overloads? Isn't this implementation the same we'd get if there were no overrides?

Reply:
Ugh - it's a record, I see. Which brings up the question -- should it be a record? Should it advertize its `size and `tl` components (they seem rather implementation specific?).

Reply:
Was this done so that the final field could be trusted? (I believe all final fields are implicitly trusted in jdk/internal/foreign):

```
      holder->is_in_package("jdk/internal/foreign/layout") || holder->is_in_package("jdk/internal/foreign") ||
```

Reply:
Yes. We should revert to a normal class.